thunderfish/sortfishtype.py

```diff
@@ -0,0 +1,189 @@
+import numpy as np
+import peakdetection as pkd
+from IPython import embed
+
+
+def type_detector(data, samplerate, thres=0.1, plot_data_func=None):
+
+    """ Detects if fish is pulse or wave by calculating the proportion of the time distance between a peak and its
+     following trough, relative to the time between consecutive peaks.
+
+    :param data: (1-D array) The data to be analyzed (Usually the best window already)
+    :param samplerate: (float). Sampling rate of the data in Hz
+    :param thres: (float or array): a positive number setting the minimum distance between peaks and troughs
+    :param plot_data_func: Function for plotting the raw data, detected peaks and troughs, the criteria,
+    the cost function and the selected best window.
+        plot_data_func(data, rate, peak_idx, trough_idx, idx0, idx1,
+                       win_start_times, cv_interv, mean_ampl, cv_ampl, clipped_frac, cost,
+                       thresh, valid_wins, **kwargs)
+        :param data (array): the raw data.
+        :param rate (float): the sampling rate of the data.
+        :param peak_idx (array): indices into raw data indicating detected peaks.
+        :param trough_idx (array): indices into raw data indicating detected troughs.
+    :return: suggestion: (str). Returns the suggested fish-type ("pulse"/"wave").
+    :return: r_value: (float). Returns a float between 0. and 1. which gives the proportion of peak-2-trough,
+                            from peak-2-peak time distance. (Wave-fishes should have larger values than pulse-fishes)
+    """
+
+    print('\nAnalyzing Fish-Type...')
+
+    # define args for peak-trough detection
+    win_shift = 0.1
+
+    # threshold for peak detection:
+    threshold = np.zeros(len(data))
+    win_shift_indices = int(win_shift * samplerate)
+
+    for inx0 in xrange(0, len(data) - win_shift_indices / 2, win_shift_indices):
+        inx1 = inx0 + win_shift_indices
+        threshold[inx0:inx1] = np.percentile(data[inx0:inx1], 99)
+
+    # detect large peaks and troughs:
+    peak_idx, trough_idx = pkd.detect_peaks(data, threshold)
+
+    # adjust array shape
+    min_len = np.min([len(peak_idx), len(trough_idx)])
+    peak_idx = peak_idx[:min_len]
+    trough_idx = trough_idx[:min_len]
+
+    # get times of peaks and troughs
+    time = np.arange(len(data)) / samplerate
+    pk_t = time[peak_idx]
+    tr_t = time[trough_idx]
+
+    pk_2_pk = np.diff(pk_t)
+    pk_2_tr = np.abs(pk_t - tr_t)
+
+    # adjust array shape
+    min_n = np.min([len(pk_2_pk), len(pk_2_tr)])
+
+    # get the proportion of peak-2-trough, from peak-2-peak time distance
+    r_tr = pk_2_tr[:min_n] / pk_2_pk[:min_n]
+    r_tr[r_tr > 0.5] = 1 - r_tr[r_tr > 0.5]
+
+    r_value = np.median(r_tr)
+    # r_value = np.median(pk_2_tr[:min_n] / pk_2_pk[:min_n])
+    suggestion = 'pulse' if r_value < thres else 'wave'
+
```

Comment:
Till and I would like this function to suggest us if fish or wave. We also thought of not merging both algorithms into one module, because the inputs are different. We'd rather have two independent modules that check if pulse- or wave-type independently. (We can discuss this)